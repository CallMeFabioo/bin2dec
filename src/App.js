import React, { useState } from 'react';

import { Input, Label, Button } from 'components';

function App() {
	const [binaryInput, setBinaryInput] = useState('');
	const [decimalOutput, setDecimalOutput] = useState('');
	const [errorMessage, setErrorMessage] = useState('');
	const [disabledButton, setToggleButton] = useState(true);

	const onChange = (e) => {
		const value = e.target.value;

		if (value.match(/^[0-1]+$/g) === null) {
			setErrorMessage('Please, enter a binary number.');
			return;
		}

		setErrorMessage('');
		setBinaryInput(value);
		setToggleButton(false);
	};

	const onSubmit = (e) => {
		e.preventDefault();

		// Formula:
		// input = 1 => output = 1 * (2^0) = 1
		// input = 10 => output = (0 * (2^0)) + (1 * (2^1)) = 2
		// So we reverse and iterate from the back
		const result = binaryInput
			.split('')
			.map(Number)
			.reverse()
			.reduce((acc, char, index) => acc + char * Math.pow(2, index));

		setDecimalOutput(result);
	};

	const clearInputs = () => {
		setBinaryInput('');
		setDecimalOutput('');
		setToggleButton(true);
	};

	return (
		<div className="flex items-center justify-center h-screen">
			<form
				className="w-1/2 px-8 pt-6 pb-8 bg-white rounded shadow-md"
				onSubmit={onSubmit}
				noValidate
			>
				<div className="mb-4">
					<Label labelFor="binaryInput">Binary input</Label>
					<Input
						id="binaryInput"
						placeholder="A 8 digit binary number"
						hasError={Boolean(errorMessage)}
						onChange={onChange}
						value={binaryInput}
					/>
					{errorMessage && (
						<p className="text-xs italic text-red-500">{errorMessage}</p>
					)}
				</div>
				<div className="mb-6">
					<Label labelFor="decimalOutput">Decimal output</Label>
					<Input
						id="decimalOutput"
						placeholder="Decimal output"
						disabled={!decimalOutput}
						value={decimalOutput}
					/>
				</div>
				<div className="flex items-center justify-around">
					<Button type="button" disabled={!decimalOutput} onClick={clearInputs}>
						Clear
					</Button>

					<Button type="submit" disabled={disabledButton}>
						Convert!
					</Button>
				</div>
			</form>
		</div>
	);
}

export default App;
