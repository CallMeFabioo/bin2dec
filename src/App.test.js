import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('Structure', () => {
	it('renders the binary input label', () => {
		const { queryByText } = render(<App />);

		const binaryLabel = queryByText(/Binary input/i);

		expect(binaryLabel).toBeInTheDocument();
	});

	it('renders the binary input', () => {
		const { getByPlaceholderText } = render(<App />);

		const binaryInput = getByPlaceholderText(/A 8 digit binary number/i);

		expect(binaryInput).toBeInTheDocument();
	});

	it('renders the decimal input label', () => {
		const { queryByText } = render(<App />);

		const decimalLabel = queryByText(/Decimal output/i);

		expect(decimalLabel).toBeInTheDocument();
	});
	it('renders the decimal input', () => {
		const { getByPlaceholderText } = render(<App />);

		const decimalInput = getByPlaceholderText(/Decimal output/i);

		expect(decimalInput).toBeInTheDocument();
	});
});

describe('Default state', () => {
	it('renders a error message if binary input has no value', () => {
		const { queryByText, getByPlaceholderText } = render(<App />);

		const binaryInput = getByPlaceholderText(/A 8 digit binary number/i);
		const errorMessage = queryByText(/Please, enter a binary number/i);

		expect(binaryInput.classList).toContain('border-red-500');
		expect(errorMessage).toBeInTheDocument();
	});

	it('renders with decimal input disabled', () => {
		const { getByPlaceholderText } = render(<App />);

		const decimalInput = getByPlaceholderText(/Decimal output/i);

		expect(decimalInput).toBeDisabled();
	});

	it('renders with "Convert!" button disabled', () => {
		const { queryByText } = render(<App />);

		const submitButton = queryByText(/Convert!/i);

		expect(submitButton).toHaveClass('disabled');
	});

	it('renders with "Clear!" button disabled', () => {
		const { queryByText } = render(<App />);

		const clearButton = queryByText(/Clear/i);

		expect(clearButton).toHaveClass('disabled');
	});
});

describe('Binary input behavior', () => {
	it('only enables the "Convert" button when the number digited is a binary digit', async () => {
		const { queryByText, getByPlaceholderText } = render(<App />);

		const clearButton = queryByText(/Clear/i);
		const submitButton = queryByText(/Convert!/i);
		const decimalInput = getByPlaceholderText(/Decimal output/i);
		const binaryInput = getByPlaceholderText(/A 8 digit binary number/i);
		const errorMessage = queryByText(/Please, enter a binary number/i);

		await userEvent.type(binaryInput, '101010');

		expect(errorMessage).not.toBeInTheDocument();
		expect(submitButton).not.toHaveClass('disabled');
		expect(clearButton).toHaveClass('disabled');
		expect(decimalInput).toBeDefined();
	});

	it('renders a error message after a non binary digit is typed', async () => {
		const { queryByText, getByPlaceholderText } = render(<App />);

		const binaryInput = getByPlaceholderText(/A 8 digit binary number/i);

		const getErrorMessage = () => queryByText(/Please, enter a binary number/i);

		let errorMessage = getErrorMessage();

		expect(errorMessage).toBeInTheDocument();

		await userEvent.type(binaryInput, '1');

		expect(errorMessage).not.toBeInTheDocument();

		userEvent.type(binaryInput, '2');

		errorMessage = getErrorMessage();

		expect(errorMessage).toBeInTheDocument();
	});
});

describe('Decimal input behavior', () => {
	it('renders a result of a conversion of binary number to decimal number', async () => {
		const { queryByText, getByPlaceholderText } = render(<App />);

		const submitButton = queryByText(/Convert!/i);
		const decimalInput = getByPlaceholderText(/Decimal output/i);
		const binaryInput = getByPlaceholderText(/A 8 digit binary number/i);

		const binaryNumber = '101010';
		const decimalNumber = '42';

		await userEvent.type(binaryInput, binaryNumber);
		await userEvent.click(submitButton);

		expect(decimalInput).toBeEnabled();
		expect(decimalInput).toHaveValue(decimalNumber);
	});
});
