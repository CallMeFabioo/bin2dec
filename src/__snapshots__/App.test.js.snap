// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render a form 1`] = `
<div>
  <div
    class="flex items-center justify-center h-screen"
  >
    <form
      class="w-1/2 px-8 pt-6 pb-8 bg-white rounded shadow-md"
      novalidate=""
    >
      <div
        class="mb-4"
      >
        <label
          class="block mb-2 text-sm text-3xl font-bold text-center text-gray-700"
          for="binaryInput"
        >
          Binary input
        </label>
        <input
          autocomplete="off"
          class="border rounded shadow w-full px-3 py-2 leading-tight text-gray-700 appearance-none focus:outline-none focus:shadow-outline"
          id="binaryInput"
          maxlength="8"
          name="binaryInput"
          placeholder="A 8 digit binary number"
          type="text"
          value=""
        />
        
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block mb-2 text-sm text-3xl font-bold text-center text-gray-700"
          for="decimalOutput"
        >
          Decimal output
        </label>
        <input
          autocomplete="off"
          class="border rounded shadow w-full px-3 py-2 leading-tight text-gray-700 appearance-none focus:outline-none focus:shadow-outline"
          disabled=""
          id="decimalOutput"
          maxlength="8"
          name="decimalOutput"
          placeholder="Decimal output"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex items-center justify-around"
      >
        <button
          class="border rounded disabled px-4 py-2 font-semibold text-blue-700 bg-transparent border-blue-500 hover:bg-blue-500 hover:text-white hover:border-transparent"
          type="button"
        >
          Clear
        </button>
        <button
          class="border rounded disabled px-4 py-2 font-semibold text-blue-700 bg-transparent border-blue-500 hover:bg-blue-500 hover:text-white hover:border-transparent"
          type="submit"
        >
          Convert!
        </button>
      </div>
    </form>
  </div>
</div>
`;
